lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs

```diff
@@ -897,59 +901,106 @@ _evaluateMinimumFee
     :: Cardano.ProtocolParameters
     -> SealedTx
     -> Maybe Coin
-_evaluateMinimumFee pp tx =
-    fromCardanoLovelace <$> minFee
+_evaluateMinimumFee pp tx = do
+    anyShelleyTx <- inAnyShelleyBasedEra (cardanoTx tx)
+    pure $ withShelleyBasedBody anyShelleyTx $ \_era body ->
+        fromCardanoLovelace $
+            Cardano.evaluateTransactionFee
+                pp
+                body
+                (estimateNumberOfWitnesses body)
+                0
+
+-- | Estimate the size of the transaction (body) when fully signed.
+_estimateSignedTransactionSize
+    :: Cardano.IsShelleyBasedEra era
+    => Cardano.ProtocolParameters
+    -> Cardano.TxBody era
+    -> TxSize
+_estimateSignedTransactionSize pparams body =
+    let
+        nWits = estimateNumberOfWitnesses body
+
+        -- Hack which allows us to rely on the ledger to calculate the size of
+        -- witnesses:
+        feeOfWits = minfee nWits - minfee 0
+        sizeOfWits =
+            case feeOfWits `quotRem` perByte of
+                (n, 0) -> TxSize n
+                (_, _) -> error $ unwords
```

Comment:
Under what circumstances will we get a `feeOfWits` that's not perfectly divisible by `perByte`?

Does the existence of a non-perfectly divisible value indicate a programming error elsewhere?

What if we were to round up to the nearest integer value, so that we can always successfully produce a `TxSize` value?

Reply:
> Under what circumstances will we get a feeOfWits that's not perfectly divisible by perByte?

There shouldn't be any, which is why I'd prefer to error out of caution. 

But could underline in the error message that this really should be impossible...